File: BOJ_백준/1916 최소비용 구하기 (다익스트라).cpp
#include <iostream>
#include <vector>
#include <queue>

#define MAX_INT 1234567890
using namespace std;

int N, M;
int start, dest, weight;
vector <pair <int, int> > city[1001];

vector <int> dijkstra(int source) {
	vector <int> dist(N + 1, MAX_INT);
	dist[source] = 0;
	priority_queue <pair <int, int> > pq;	// 정렬해야하니까 
											// cost, vertex
	pq.push({ dist[source], source });

	int cost, vertex, nb, nb_dist;
	while (!pq.empty()) {
		// 현재에서 갈 수 있는 최단거리
		cost = -pq.top().first;
		vertex = pq.top().second;
		pq.pop();

		for (int i = 0; i < city[vertex].size(); i++) {
			nb = city[vertex][i].first;
			nb_dist = cost + city[vertex][i].second;

			//최소 경로 발견
			if (dist[nb] > nb_dist) {
				dist[nb] = nb_dist;
				pq.push({ -nb_dist, nb });
			}
		}
	}
	return dist;
}

int main() {
	ios_base::sync_with_stdio(false);
	cin.tie(NULL);
	cin >> N >> M;

	for (int i = 0; i < M; i++) {
		cin >> start >> dest >> weight;
		city[start].push_back({ dest, weight });
	}
	cin >> start >> dest;

	vector <int> result = dijkstra(start);
	cout << result[dest];
	return 0;
}

File: BOJ_백준/1916 최소비용 구하기 (다익스트라).py
import sys
import heapq

input = sys.stdin.readline

def dijkstra(source: int) :
    dist = [INF] * (N + 1)
    dist[source] = 0

    pq = []
    heapq.heappush(pq, [dist[source], source])

    while pq:
        cost, vertex = heapq.heappop(pq)
        
        for nb, nb_cost in city[vertex]:
            nb_dist = nb_cost + cost
            
            if dist[nb] > nb_dist :
                dist[nb] = nb_dist
                heapq.heappush(pq, [nb_dist, nb])
    return dist


INF = 1e9
N = int(input())
M = int(input())

city = [[] for _ in range(N + 1)]
for i in range(M):
    s, d, w = map(int, input().split())
    city[s].append([d, w])

start, dest = map(int, input().split())

result = dijkstra(start)
print(result[dest])
